c_sources/osqp_mex.cpp

```diff
@@ -0,0 +1,644 @@
+        OSQPSettingsWrapper settings;
+        plhs[0] = settings.GetMxStruct();
+        return;
+    }
+
+    // Return solver constants
+    if (!strcmp("constant", cmd)) {
+        static std::map<std::string, OSQPFloat> floatConstants{
+            // Numerical constants
+            {"OSQP_INFTY", OSQP_INFTY}
+        };
+
+        static std::map<std::string, OSQPInt> intConstants{
+            // Return codes
+            {"OSQP_SOLVED",                       OSQP_SOLVED},
+            {"OSQP_SOLVED_INACCURATE",            OSQP_SOLVED_INACCURATE},
+            {"OSQP_UNSOLVED",                     OSQP_UNSOLVED},
+            {"OSQP_PRIMAL_INFEASIBLE",            OSQP_PRIMAL_INFEASIBLE},
+            {"OSQP_PRIMAL_INFEASIBLE_INACCURATE", OSQP_PRIMAL_INFEASIBLE_INACCURATE},
+            {"OSQP_DUAL_INFEASIBLE",              OSQP_DUAL_INFEASIBLE},
+            {"OSQP_DUAL_INFEASIBLE_INACCURATE",   OSQP_DUAL_INFEASIBLE_INACCURATE},
+            {"OSQP_MAX_ITER_REACHED",             OSQP_MAX_ITER_REACHED},
+            {"OSQP_NON_CVX",                      OSQP_NON_CVX},
+            {"OSQP_TIME_LIMIT_REACHED",           OSQP_TIME_LIMIT_REACHED},
+
+            // Linear system solvers
+            {"QDLDL_SOLVER",         QDLDL_SOLVER},
+            {"OSQP_UNKNOWN_SOLVER",  OSQP_UNKNOWN_SOLVER},
+            {"OSQP_DIRECT_SOLVER",   OSQP_DIRECT_SOLVER},
+            {"OSQP_INDIRECT_SOLVER", OSQP_INDIRECT_SOLVER}
+        };
+
+        char constant[64];
+        int  constantLength = mxGetN(prhs[1]) + 1;
+        setString(nrhs, prhs, 1, constant, constantLength);
+
+        auto ci = intConstants.find(constant);
+
+        if(ci != intConstants.end()) {
+            plhs[0] = mxCreateDoubleScalar(ci->second);
+            return;
+        }
+
+        auto cf = floatConstants.find(constant);
+
+        if(cf != floatConstants.end()) {
+            plhs[0] = mxCreateDoubleScalar(cf->second);
+            return;
+        }
+
+        // NaN is special because we need the Matlab version
+        if (!strcmp("OSQP_NAN", constant)){
+            plhs[0] = mxCreateDoubleScalar(mxGetNaN());
+            return;
+        }
+
+        mexErrMsgTxt("Constant not recognized.");
+
+        return;
+    }
+
+    /*
+     * Finally, check to see if this is a function operating on a solver instance
+     */
+
+    // Check for a second input, which should be the class instance handle
+    if (nrhs < 2) mexErrMsgTxt("Second input should be a class instance handle.");
+
+    // Get the class instance pointer from the second input
+    osqpData = convertMat2Ptr<OsqpData>(prhs[1]);
+
+    // delete the object and its data
+    if (!strcmp("delete", cmd)) {
+        osqp_cleanup(osqpData->solver);
+        destroyObject<OsqpData>(prhs[1]);
+        // Warn if other commands were ignored
+        if (nlhs != 0 || nrhs != 2)
+            mexWarnMsgTxt("Delete: Unexpected arguments ignored.");
+        return;
+    }
```

Comment:
I'm a bit confused about this. During the constructor ("new"), I dynamically create a new OSQPCodegenDefines object (see Line 159). Seeing how osqp_cleanup destroys osqpData->solver, I thought I would have to free osqpData->defines manually. However, when I did that, the program crashed because destroyObject tries to destroy the already freed memory. What I don't understand is why this doesn't happen with osqpData->solver, which is freed by osqp_cleanup.